app_dart/lib/src/request_handlers/update_task_status.dart

```diff
@@ -111,6 +117,58 @@ class UpdateTaskStatus extends ApiRequestHandler<UpdateTaskStatusResponse> {
     return UpdateTaskStatusResponse(task);
   }
 
+  /// Retrieve [Task] to update from [DatastoreService].
+  Future<Task> _getTask(DatastoreService datastore) async {
+    if (requestData.containsKey(taskKeyParam)) {
+      return _getTaskFromEncodedKey(datastore);
+    }
+
+    return _getTaskFromNamedParams(datastore);
+  }
+
+  /// Retrieve [Task] from [DatastoreService] when given [taskKeyParam].
+  ///
+  /// This is used for Devicelab test runs from Cocoon agents. The Cocoon agent is scheduled tasks
+  /// from the Cocoon backend and is aware of the Datastore task key.
+  ///
+  // TODO(chillers): Remove this when Devicelab is migrated to LUCI. https://github.com/flutter/flutter/projects/151
+  Future<Task> _getTaskFromEncodedKey(DatastoreService datastore) {
+    final ClientContext clientContext = authContext.clientContext;
+    final KeyHelper keyHelper = KeyHelper(applicationContext: clientContext.applicationContext);
+    final Key taskKey = keyHelper.decode(requestData[taskKeyParam] as String);
+    return datastore.db.lookupValue<Task>(taskKey, orElse: () {
+      throw BadRequestException('No such task: ${taskKey.id}');
+    });
+  }
+
+  /// Retrieve [Task] from [DatastoreService] when given [gitShaParam], [gitBranchParam], and [taskNameParam].
+  ///
+  /// This is used when the DeviceLab test runner is uploading results to Cocoon for runs on LUCI.
+  /// LUCI does not know the [Key] assigned to task when scheduling the build, but Cocoon can
+  /// lookup the task based on these key values.
+  ///
+  /// To lookup the value, we construct the ancestor key, which corresponds to the [Commit].
+  /// Then we query the tasks with that ancestor key and filter for the one that has the corresponding name.
+  Future<Task> _getTaskFromNamedParams(DatastoreService datastore) async {
```

Comment:
add checkRequiredParameters for [gitShaParam], [gitBranchParam], [taskNameParam] when the task is from luci.

Reply:
Added check for this when from luci, otherwise verify taskKeyParam was passed.